chain/chain/src/chain.rs

```diff
@@ -1038,6 +1041,94 @@ impl Chain {
         }
     }
 
+    /// Check if any orphan with missing chunks is ready to request chunks
+    pub fn check_orphans_with_missing_chunks<F>(
+        &mut self,
+        me: &Option<AccountId>,
+        mut orphans_missing_chunks: F,
+    ) where
+        F: Copy + FnMut((CryptoHash, Vec<ShardChunkHeader>)) -> (),
+    {
+        let mut orphan_hashes = vec![];
+        for (orphan_hash, orphan) in self.orphans.orphans.iter() {
+            let mut accepted_parent = false;
+            let mut accepted_parent_hash = CryptoHash::default();
+            let mut all_parents_known = true;
+            let mut prev_hash = orphan.block.header.prev_hash;
+            for _ in 0..NUM_ORPHAN_PARENTS_CHECK {
+                if prev_hash == CryptoHash::default() {
+                    // TODO This is very suspicious.
+                    // Technically, it's correct and also helps in testing.
+                    // In reality it show never happen except maybe first seconds
+                    // of creating Universe.
+                    break;
+                }
+                // 1. Look into store
+                match self.store.get_block(&prev_hash) {
+                    Ok(block) => {
+                        accepted_parent = true;
+                        accepted_parent_hash = prev_hash;
+                        prev_hash = block.header.prev_hash;
+                        continue;
+                    }
+                    _ => {}
+                }
+                // 2. Look into missing chunks
+                if self.blocks_with_missing_chunks.orphans.contains_key(&prev_hash) {
+                    let prev_orphan =
+                        self.blocks_with_missing_chunks.orphans.get(&prev_hash).unwrap();
+                    prev_hash = prev_orphan.block.header.prev_hash;
+                    continue;
+                }
+                // 3. Look into orphans
+                if self.orphans.orphans.contains_key(&prev_hash) {
+                    let prev_orphan = self.orphans.orphans.get(&prev_hash).unwrap();
+                    prev_hash = prev_orphan.block.header.prev_hash;
+                    continue;
+                }
+                // Parent is not found
+                all_parents_known = false;
+                break;
+            }
+            // The following conditions should be fulfilled to request chunks:
+            // 1. At least one of the parents is accepted (is in store)
+            // 2. All of the parents_to_check should be known (in store, missing chunks or orphans)
+            if accepted_parent && all_parents_known {
+                orphan_hashes.push((accepted_parent_hash, orphan_hash));
+            }
+        }
+        if self.orphans.orphans.len() > 0 {
+            debug!(target: "chain", "Orphans total: {:?}, orphans to get chunks: {:?}", self.orphans.orphans.len(), orphan_hashes.len());
+        }
+
+        let mut chain_update = ChainUpdate::new(
```

Comment:
Extract this into a method, and replace all occurrences in the code base :)

Reply:
Is it possible?
We mutable borrowing chain.store here, it means the code in the function knows that we consumes store only and allows us to borrow other fields.
In case of covering into `ChainUpdate::new(self)` or `self.gimme_chain_update()` the compiler expects we mutable borrows all fields of chain which makes the method useless.
`ChainUpdate::new(self, &mut self.store)` can't compile because of mixing immutable and mutable borrowing.
And, finally, `self.store.gimme_chain_update()` doesn't work because store at least doesn't know about ChainUpdate and ChainStoreUpdate.

Technically, we want to move the `store` field and guarantee we will not use in further. I think it's fine in C++, not in Rust.
Any ideas?